jax/_src/scipy/spatial/transform.py

```diff
@@ -0,0 +1,440 @@
+# Copyright 2023 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import re
+import typing
+
+import jax
+import jax.numpy as jnp
+
+
+class Rotation(typing.NamedTuple):
+  """Rotation in 3 dimensions."""
+
+  quat: jax.Array
+
+  @classmethod
+  def concatenate(cls, rotations: typing.Sequence):
+    """Concatenate a sequence of `Rotation` objects."""
+    return cls(jnp.concatenate([rotation.quat for rotation in rotations]))
+
+  @classmethod
+  def from_euler(cls, seq: str, angles: jax.Array, degrees: bool = False):
+    """Initialize from Euler angles."""
+    assert angles.ndim in [1, 2]
+    num_axes = len(seq)
+    if num_axes < 1 or num_axes > 3:
+      raise ValueError("Expected axis specification to be a non-empty "
+                       "string of upto 3 characters, got {}".format(seq))
+    intrinsic = (re.match(r'^[XYZ]{1,3}$', seq) is not None)
+    extrinsic = (re.match(r'^[xyz]{1,3}$', seq) is not None)
+    if not (intrinsic or extrinsic):
+      raise ValueError("Expected axes from `seq` to be from ['x', 'y', "
+                       "'z'] or ['X', 'Y', 'Z'], got {}".format(seq))
+    if any(seq[i] == seq[i+1] for i in range(num_axes - 1)):
+      raise ValueError("Expected consecutive axes to be different, "
+                       "got {}".format(seq))
+    seq = seq.lower()
+    _from_euler = lambda a: _elementary_quat_compose(a, seq, intrinsic, degrees)
+    if angles.ndim == 1:
+      return cls(_from_euler(angles))
+    else:
+      return cls(jax.vmap(_from_euler)(angles))
```

Comment:
In many of these cases where we choose to `vmap` depending on the input shape, it might be cleaner to wrap the source function in `jnp.vectorize` with an appropriate signature.

This would allow more general batch sizes, and would simplify the implementation here. What do you think?

Reply:
Wow, this is great! Much simpler and more powerful.